kmod/src/data.c

```diff
@@ -1889,6 +1889,77 @@ int scoutfs_data_waiting(struct super_block *sb, u64 ino, u64 iblock,
 	return ret;
 }
 
+static int scoutfs_data_filemap_fault(struct vm_area_struct *vma, struct vm_fault *vmf)
+{
+	struct file *file = vma->vm_file;
+	struct inode *inode = file_inode(file);
+	struct scoutfs_inode_info *si = SCOUTFS_I(inode);
+	struct super_block *sb = inode->i_sb;
+	struct scoutfs_lock *inode_lock = NULL;
+	SCOUTFS_DECLARE_PER_TASK_ENTRY(pt_ent);
+	DECLARE_DATA_WAIT(dw);
+	bool have_ret = false;
+	loff_t pos;
+	int err;
+	vm_fault_t ret;
+
+	pos = vmf->pgoff;
+	pos <<= PAGE_SHIFT;
+
+retry:
+	err = scoutfs_lock_inode(sb, SCOUTFS_LOCK_READ,
+				 SCOUTFS_LKF_REFRESH_INODE, inode, &inode_lock);
+	if (err < 0)
+		return VM_FAULT_SIGBUS;
+
+	if (scoutfs_per_task_add_excl(&si->pt_data_lock, &pt_ent, inode_lock)) {
+		/* protect checked extents from stage/release */
+		atomic_inc(&inode->i_dio_count);
+
+		err = scoutfs_data_wait_check(inode, pos, PAGE_SIZE,
+					      SEF_OFFLINE, SCOUTFS_IOC_DWO_READ,
+					      &dw, inode_lock);
+		if (err != 0)
+			goto out;
+	}
+
+	ret = filemap_fault(vma, vmf);
+	have_ret = true;
+
+out:
+	if (scoutfs_per_task_del(&si->pt_data_lock, &pt_ent))
+		inode_dio_done(inode);
+	scoutfs_unlock(sb, inode_lock, SCOUTFS_LOCK_READ);
+	if (scoutfs_data_wait_found(&dw)) {
+		int err = scoutfs_data_wait(inode, &dw);
```

Comment:
There's a lot of confusing stuff around the way this calls data_wait (you can't have have_ret and _data_wait_found both be true.  But honestly, I'm not sure if it needs to do any of this at all.  filemap_fault will call ->readpage which will wait for the data.  So tear all the waiting stuff out from here and write a quick test to use xfs_io to do an mmap read from an offline file and see if it triggers staging and gets the correct data?

Reply:
So, I believe that this code is very much still needed. I've modified the `tests/mmap.sh` script in the next revision to offline an extent and then test `mread` and `mwrite` on that using xfs_io, and with the data wait stuff in this function, every thing is just happy (Aside the have_ret confusion).

If I remove any of the data wait stuff, it all falls over and mread() returns all zeroes instead of the expected pattern - clearly there's something at hand here that this is still needed.

The next iteration of this code will have the `have_ret` stuff removed out of this particular code path because indeed that's not right the way it is, and that holds up nicely in testing.